### PR TITLE
TSQL: `CREATE EXTERNAL FILE FORMAT`

### LINKED ISSUE #4610: TSQL: `CREATE EXTERNAL FILE FORMAT` unparsable
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Microsoft SQL Server 2016 and later supports `CREATE EXTERNAL FILE FORMAT` statement:
https://learn.microsoft.com/en-us/sql/t-sql/statements/create-external-file-format-transact-sql

This example from the [official TSQL documentation](https://learn.microsoft.com/en-us/sql/t-sql/statements/create-external-file-format-transact-sql?view=sql-server-ver16&tabs=delimited#a-create-a-delimitedtext-external-file-format) is not parsable:
```sql
CREATE EXTERNAL FILE FORMAT textdelimited1
WITH (
    FORMAT_TYPE = DELIMITEDTEXT,
    FORMAT_OPTIONS (
        FIELD_TERMINATOR = '|',
        DATE_FORMAT = 'MM/dd/yyyy' ),
    DATA_COMPRESSION = 'org.apache.hadoop.io.compress.GzipCodec'
);
```

### Expected Behaviour

Above example should be parsable.

### Observed Behaviour

`Found unparsable section`

### How to reproduce

[see example above](https://online.sqlfluff.com/fluffed?sql=H4sIAB2yHWQC_1WOzYrCMBSF94W-w91VoURmLS5Ce50JNE1JL8zMKpQmjIWRFO1iFB_eq1ZkzvZ856ewKAkBvwhtLSvYqgpha6yWBFP4m3z4HfbDFPxbmnwq-oBFmgDrgTj6bhA2UGKltCIsiYvyf4RpSJm6febujsKqdDyoVS3JWC7ILln-Akq-5OYT7Gm98n51YmWwnDFGpCuMbiy2LQ_cuHj4Ed3Y9bsgdp2PcRRDFH3cj4dwPIr38zAW0Yc-S5Pl-goApDOr-AAAAA%3D%3D&dialect=tsql)

### Dialect

tsql

### Version

2.0.2 & webversion

### Configuration

not applicable

### Are you willing to work on and submit a PR to address the issue?

- [x] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by keen85**
### Brief summary of the change made
fixes #4610

### Are there any other side effects of this change that we should be aware of?
No

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/dialects/dialect_tsql.py [hunk 1]
```diff
@@ -5140,3 +5190,244 @@ class CreateExternalDataSourceStatementSegment(BaseSegment):
...
+    """
+
+    type = "external_file_orc_clause"
+
+    match_grammar = Delimited(
+        Sequence(
+            "FORMAT_TYPE",
+            Ref("EqualsSegment"),
+            "ORC",
+            optional=False,
```
Comment (WittierDinosaur) [inline src/sqlfluff/dialects/dialect_tsql.py:5328]: optional defaults to false

File: src/sqlfluff/dialects/dialect_tsql.py [hunk 2]
```diff
@@ -5140,3 +5190,244 @@ class CreateExternalDataSourceStatementSegment(BaseSegment):
...
+        ),
+        Sequence(
+            "DATA_COMPRESSION",
+            Ref("EqualsSegment"),
+            OneOf(
```
Comment (WittierDinosaur) [inline src/sqlfluff/dialects/dialect_tsql.py:5333]: this OneOf isn't needed

File: src/sqlfluff/dialects/dialect_tsql.py [hunk 3]
```diff
@@ -5140,3 +5190,244 @@ class CreateExternalDataSourceStatementSegment(BaseSegment):
...
+        Sequence(
+            "FORMAT_TYPE",
+            Ref("EqualsSegment"),
+            "PARQUET",
+            optional=False,
+        ),
+        Sequence(
+            "DATA_COMPRESSION",
+            Ref("EqualsSegment"),
+            OneOf(
```
Comment (WittierDinosaur) [inline src/sqlfluff/dialects/dialect_tsql.py:5359]: this OneOf isn't needed

File: src/sqlfluff/dialects/dialect_tsql.py [hunk 4]
```diff
@@ -5140,3 +5190,244 @@ class CreateExternalDataSourceStatementSegment(BaseSegment):
...
+        Sequence(
+            "FORMAT_TYPE",
+            Ref("EqualsSegment"),
+            "JSON",
+            optional=False,
+        ),
+        Sequence(
+            "DATA_COMPRESSION",
+            Ref("EqualsSegment"),
+            OneOf(
```
Comment (WittierDinosaur) [inline src/sqlfluff/dialects/dialect_tsql.py:5385]: This OneOf isn't needed

File: src/sqlfluff/dialects/dialect_tsql.py [hunk 5]
```diff
@@ -5140,3 +5190,244 @@ class CreateExternalDataSourceStatementSegment(BaseSegment):
...
     )
+
+
+class ExternalFileFormatDelimitedTextFormatOptionClause(BaseSegment):
+    """`CREATE EXTERNAL FILE FORMAT` Delimited text `FORMAT_OPTIONS` clause."""
+
+    type = "external_file_delimited_text_format_options_clause"
+
+    match_grammar = OneOf(
+        Sequence(
```
Comment (WittierDinosaur) [inline src/sqlfluff/dialects/dialect_tsql.py:5201]: Most of these can be condensed into one Sequence. The ones that equal quotedliteral can have the options wrapped in a OneOf()
OneOf("FIELD_TERMINATOR", "STRING_DELIMITER") etc

**Review by WittierDinosaur — CHANGES_REQUESTED** [review]
Left a few comments. There are more places where there are similary unnecessary OneOf's and optional=False

**Comment by keen85** [comment]
@WittierDinosaur thanks for the valuable input.

File: src/sqlfluff/dialects/dialect_tsql.py [hunk 6]
```diff
@@ -5162,3 +5210,209 @@ class PeriodSegment(BaseSegment):
...
+            "FORMAT_TYPE",
+            Ref("EqualsSegment"),
+            "DELIMITEDTEXT",
+        ),
+        Sequence(
+            "FORMAT_OPTIONS",
+            Bracketed(
+                Delimited(
+                    AnyNumberOf(
```
Comment (WittierDinosaur) [inline src/sqlfluff/dialects/dialect_tsql.py:5264]: Last one - you don't need this AnyNumberOf

**Review by WittierDinosaur — APPROVED** [review]
LGTM

**Pull request merged**